File: packages/backup/backup.js
let commandExists = require("command-exists");
const { exec } = require("child_process");

function executeCommand(command) {
  const exec = require('child_process').exec;

  exec(command, (err, stdout, stderr) => {
    process.stdout.write(stdout);
  })
}

const SM_GH_UN = () => {
  executeCommand(
    "git config user.name"
  );
};

const create = () => {
  executeCommand(
    `gh repo create ${SM_GH_UN}/.secman.bk -y --private`
  );
};

// dirs
let SECDIR = "~/.secman.bk";
let SECDIR_primary = "~/.secman";

const cd_SECDIR = () => {
  executeCommand(
    `cd ${SECDIR}`
  );
};

// pkgs
const ghraw_url = "https://raw.githubusercontent.com";

let install_cgit = () => {
    executeCommand(`/bin/bash -c \"$(curl -fsSL ${ghraw_url}/Dev-x-Team/corgit/main/setup)\"`)
}

let install_brew = () => {
    executeCommand(`/bin/bash -c \"$(curl -fsSL ${ghraw_url}/Homebrew/install/HEAD/install.sh)\"`)
}

let install_verx = () => {
    executeCommand(`/bin/bash -c \"$(curl -fsSL ${ghraw_url}/abdfnx/verx/HEAD/install.sh)\"`)
}

const brew_gh = () => {
  executeCommand(
    "brew install gh"
  );
};

function _help() {
    const u =
        "Flags:\n\t-h | --help: help about any command\nCommands:\n\tsy | sync: create private github repo and sync your passwords on it by git\n\tcn | clone: clone .secman repo\n\tph | push: push and commit a new secret\n\tpl | pull: pull secret/s";

    console.log(u);
}

const csi = () => {
    executeCommand("cgit secman-ibk");
}

const rdm = () => {
    executeCommand(`echo \"# My secman backup passwords - ${SM_GH_UN}\" >> ${SECDIR}/README.md`);
}

const mkdir = () => {
    executeCommand(`cp -r ~/.secman ~/.secman.bk`)
}

function repo_work () {
    mkdir();
    rdm();
    create();
    executeCommand(
        `cp -r ~/.secman ~/.secman.bk && gh repo create ${SM_GH_UN}/.secman.bk -y --private && cd ~/.secman.bk && cgit secman-ibk`
    );
}

repo_work();

// function repo() {
//     commandExists("gh", function (err, commandExists) {
//         if (commandExists) {
//             repo_work();
//         } else {
//             commandExists("brew", function (err, commandExists) {
//                 if (commandExists) {
//                     brew_gh();

//                     commandExists("gh", function (err, commandExists) {
//                         if (commandExists) {
//                             repo_work();
//                         }
//                     });
//                 } else {
//                     install_brew();

//                     commandExists("brew", function (err, commandExists) {
//                         if (commandExists) {
//                             brew_gh();

//                             commandExists("gh", function (err, commandExists) {
//                                 if (commandExists) {
//                                     repo_work();
//                                 }
//                             });
//                         }
//                     });
//                 }
//             });
//         }
//     });
// }

// const _ph = () => {
//     const push = () => {
//         executeCommand("cgit ph");
//     }

//     commandExists("cgit", function (err, commandExists) {
//         if (commandExists) {
//             cd_SECDIR();
//             push();
//         } else {
//             install_cgit();

//             if (commandExists) {
//                 cd_SECDIR();
//                 push();
//             }
//         }
//     });
// }

// const _pl = () => {
//     const pull = () => {
//         executeCommand("cgit pl");
//     }

//     commandExists("cgit", function (err, commandExists) {
//         if (commandExists) {
//             cd_SECDIR();
//             pull();
//         } else {
//             install_cgit();

//             if (commandExists) {
//                 cd_SECDIR();
//                 pull();
//             }
//         }
//     });
// }

// function _clone() {
//     const clone = () => {
//         `gh repo clone ${SM_GH_UN}/.secman.bk ${SECDIR}`
//     }

//     clone();
// }

// const version = () => {
//     let v = () => {
//         executeCommand("secman ver");
//     }

//     commandExists("verx", function (err, commandExists) {
//         if (commandExists) {
//             v();
//         } else {
//             install_verx();

//             if (commandExists) {
//                 v();
//             }
//         }
//     });
// }

// var menuHandler;

// // Initialize
// function initialize() {
//     showMenu();

//     process.stdin.setEncoding('utf8');
//     process.stdin.on('readable', checkMenu);

//     function checkMenu() {
//         var input = process.stdin.read();
//         if(input !== null) {
//             menuHandler(input.trim());
//         }
//     }
// }

// // Main
// function showMenu() {
//     console.log(
//         'Init backup (init)' + '\n' +
//         'Clone backup (clone)'  + '\n' +
//         'Push the new passwords in backup (push)'  + '\n' +
//         'Pull (pull)'  + '\n\n' +
//         'Choose operation, then press ENTER: '
//     );

//     menuHandler = function(input) {
//         switch(input) {
//             case "init":
//                 repo();
//                 break;
            
//             case "clone":
//                 _clone();
//                 break;
            
//             case "push":
//                 _ph();
//                 break;
            
//             case "pull":
//                 _pl();
//                 break;
            
//             default:
//                 _help();
//                 break;
//         }
//     };
// }

// initialize();

File: packages/backup/backup.py
import os
import sys, getopt
import subprocess as sp
import pathlib
import platform

# dirs
SECDIR = "~/.secman.bk"
SECDIR_primary = "~/.secman"
SECDIR_url = "secman.bk"
SECDIR_primary_url = ".secman"
cd_SECDIR = "cd {}".format(SECDIR)

# git & github
SM_GH_UN = sp.getoutput("git config user.name")
create = "gh repo create {}/{} -y --private".format(SM_GH_UN, SECDIR_url)
clone = "gh repo clone {}/{} {}".format(SM_GH_UN, SECDIR_url, SECDIR)
remote = "git remote set-url origin https://github.com/{}/.secman".format(
    SM_GH_UN)

# pkgs
ghraw_url = "https://raw.githubusercontent.com"

def install(_url):
    return "curl -fsSL {}/{} | bash".format(ghraw_url, _url)

install_brew = install("Homebrew/install/HEAD/install.sh")

brew_gh = "brew install gh"
choco_gh = "choco install gh"

def _help():
    print(
        "Flags:\n\t-h: help\n\t-i: init ~/.secman.bk folder\n\t-c: clone the backup repo\n\t-p: push new backup passwords\n\t-l: pull passwords\n\t-M: make ~/.secman.bk is the main folder\n"
    )

def repo_work():
    csi = "bash ~/sm/cgit secman-ibk"
    rdm = 'touch {}/README.md && echo "# My secman backup passwords - {}" >> {}/README.md'.format(
        SECDIR, SM_GH_UN, SECDIR)

    # copy ~/.secman to ~/.secman.bk
    sp.getoutput("cp -rf {} {}".format(SECDIR_primary, SECDIR))
    os.system("{} && rm -rf .git && git init && {} && {} && {}".format(
        cd_SECDIR, rdm, create, csi))

def install_installer():
    if platform.system() == "Linux" or platform.system() == "Darwin":
        sp.getoutput(brew_gh)
    elif platform.system() == "Windows":
        sp.getoutput(choco_gh)

def repo():
    try:
        devnull = open(os.devnull)
        sp.Popen(["gh"], stdout=devnull, stderr=devnull).communicate()
        repo_work()
    except OSError:
        install_installer()

        try:
            devnull = open(os.devnull)
            sp.Popen(["gh"], stdout=devnull, stderr=devnull).communicate()
            repo_work()

        except OSError:
            if platform.system() == "Linux" or platform.system() == "Darwin":
                sp.getoutput(install_brew)
            elif platform.system() == "Windows":
                print(
                    "you should install Chocolatey from this url https://chocolatey.org/install"
                )

            try:
                devnull = open(os.devnull)

                if platform.system() == "Linux" or platform.system(
                ) == "Darwin":
                    sp.Popen(["brew"], stdout=devnull,
                             stderr=devnull).communicate()
                elif platform.system() == "Windows":
                    sp.Popen(["choco"], stdout=devnull,
                             stderr=devnull).communicate()

                install_installer()

                try:
                    devnull = open(os.devnull)
                    sp.Popen(["gh"], stdout=devnull,
                             stderr=devnull).communicate()
                    repo_work()

                except OSError:
                    print(
                        "sorry, there's an error while initialize a backup, try again"
                    )

            except OSError:
                if platform.system() == "Linux" or platform.system(
                ) == "Darwin":
                    print("you should install brew\nhttps://brew.sh")
                elif platform.system() == "Windows":
                    print(
                        "you should install Chocolatey from this url https://chocolatey.org/install"
                    )
        return False

    return True

def _ph():
    os.system("{} && bash ~/sm/cgit ph".format(cd_SECDIR))

def _pl():
    os.system("{} && bash ~/sm/cgit pl".format(cd_SECDIR))

def _clone():
    os.system(clone)

def badUsage():
    print(
        "Flag not recognized.\nFor an overview of the command, execute: secman-sync backup -h"
    )

def version():
    os.system("secman ver")

def make_main():
    """
    1. check if ~/.secman is not exist
     then
        1. copy ~/.secman.bk to ~/.secman
        2. change git remote

    2. if ~/.secman is exist
     then
        execute this msg '~/.secman is exist'
    """

    home = os.path.expanduser('~/.secman')
    SECDIR_path = pathlib.Path(home)

    if not SECDIR_path.exists():
        os.system("cp -rf {} {}".format(SECDIR, SECDIR_primary))
        os.system("cd {} && {}".format(SECDIR_primary, remote))

    else:
        print("{} is exist".format(SECDIR_primary))

def main(argv):
    try:
        opts, args = getopt.getopt(
            argv, "hicvplM", ["help", "init", "clone", "push", "pull", "Main"])

    except getopt.GetoptError:
        badUsage()
        sys.exit(2)

    for opt, arg in opts:
        if opt in ("-h", "--help"):
            _help()
            sys.exit()

        elif opt in ("-i", "--init"):
            repo_work()
            sys.exit()

        elif opt in ("-c", "--clone"):
            _clone()
            sys.exit()

        elif opt in ("-v", "--version"):
            version()
            sys.exit()

        elif opt in ("-p", "--push"):
            _ph()
            sys.exit()

        elif opt in ("-l", "--pull"):
            _pl()
            sys.exit()

        elif opt in ("-m", "-M", "--Main"):
            make_main()
            sys.exit()

        else:
            badUsage()
            sys.exit()

if __name__ == "__main__":
    main(sys.argv[1:])
